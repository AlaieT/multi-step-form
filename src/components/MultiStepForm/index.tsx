import React, { useMemo, useState } from "react";
import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";

import Steps from "../Steps";
import { multiStepFromSchema } from "../../utils/schemas";
import { FormContext } from "../../utils/context";

import type { MultiStepFormProps, MFS, FormContextType } from "../../types";

import styles from "../../styles/components/multiStepFrom.module.scss";

const MultiStepForm = ({
  steps: { pickAddOns, selectYourPlan },
  onSubmit
}: MultiStepFormProps) => {
  const [step, setStep] = useState<number>(0);
  const [form, setForm] = useState<FormContextType["form"]>({
    planMode: "yearly",
    priceMode: "yr"
  });
  const formContextValue = useMemo<FormContextType>(
    () => ({ form, setForm }),
    [form, setForm]
  );
  const {
    register,
    getValues,
    setValue,
    trigger,
    formState: {
      errors, isValid, isSubmitted, isSubmitSuccessful
    },
    handleSubmit
  } = useForm<MFS>({
    defaultValues: {
      name: "",
      email: "",
      phone: "",
      plan: "arcade",
      planMode: true,
      customizableProfile: false,
      onlineService: false,
      largeStorage: false,
      totalPrice: 0
    },
    resolver: yupResolver(multiStepFromSchema),
    mode: "onChange"
  });

  const handleChangeStep = (newStep: number) => {
    if (newStep >= 0 && newStep < 4 && !isSubmitted && !isSubmitSuccessful) {
      setStep(newStep);
    }
  };

  return (
    <FormContext.Provider value={formContextValue}>
      <div id={styles.box}>
        <div id={styles.steps}>
          <nav>
            <ul>
              <li id={step === 0 ? styles.active : undefined}>
                <p>STEP 1</p>
                <h3>YOUR INFO</h3>
              </li>
              <li id={step === 1 ? styles.active : undefined}>
                <p>STEP 2</p>
                <h3>SELECT PLAN</h3>
              </li>
              <li id={step === 2 ? styles.active : undefined}>
                <p>STEP 3</p>
                <h3>ADD-ONS</h3>
              </li>
              <li id={step === 3 ? styles.active : undefined}>
                <p>STEP 4</p>
                <h3>SUMMARY</h3>
              </li>
            </ul>
          </nav>
        </div>
        <form
          aria-label="form"
          onSubmit={handleSubmit((data) => onSubmit(data))}
        >
          {step === 0 && !isSubmitted && !isSubmitSuccessful && (
            <Steps.YourInfo
              trigger={trigger}
              register={register}
              getValues={getValues}
              errors={errors}
              changeStep={handleChangeStep}
            />
          )}
          {step === 1 && !isSubmitted && !isSubmitSuccessful && (
            <Steps.SelectYourPlan
              plans={selectYourPlan.plans}
              register={register}
              getValues={getValues}
              changeStep={handleChangeStep}
            />
          )}
          {step === 2 && !isSubmitted && !isSubmitSuccessful && (
            <Steps.PickAddOns
              addOns={pickAddOns.addOns}
              register={register}
              getValues={getValues}
              changeStep={handleChangeStep}
            />
          )}
          {step === 3 && !isSubmitted && !isSubmitSuccessful && (
            <Steps.Summary
              plans={selectYourPlan.plans}
              addOns={pickAddOns.addOns}
              isValid={isValid}
              setValue={setValue}
              getValues={getValues}
              changeStep={handleChangeStep}
            />
          )}
          {isSubmitted && isSubmitSuccessful && <Steps.Finish />}
        </form>
      </div>
    </FormContext.Provider>
  );
};

export default MultiStepForm;
