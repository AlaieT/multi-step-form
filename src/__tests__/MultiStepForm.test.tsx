import React from "react";
import {
  render,
  cleanup,
  screen,
  fireEvent,
  act
} from "@testing-library/react";

import MultiStepForm from "../components/MultiStepForm";

afterEach(cleanup);

describe("MultiStepForm", () => {
  const props = {
    selectYourPlan: { plans: global.getStepProps().plans },
    pickAddOns: { addOns: global.getStepProps().addOns }
  };
  let formData = null;
  const mockOnSubmit = (data) => (formData = data);

  describe("correct render", () => {
    it("should match snapshot for all steps", async () => {
      /**YourInfo step */
      const { rerender, asFragment } = render(
        <MultiStepForm onSubmit={mockOnSubmit} steps={props} />
      );

      expect(asFragment()).toMatchSnapshot();

      fireEvent.input(screen.getByRole("textbox", { name: "Name" }), {
        target: { value: "Some name" }
      });
      fireEvent.input(screen.getByRole("textbox", { name: "Email Address" }), {
        target: { value: "some@mail.dom" }
      });
      fireEvent.input(screen.getByRole("textbox", { name: "Phone Number" }), {
        target: { value: "Some address" }
      });

      /**SelectYourPlan step */
      await act(() => {
        fireEvent.click(screen.getByRole("button"));
      });

      rerender(<MultiStepForm onSubmit={mockOnSubmit} steps={props} />);

      expect(asFragment()).toMatchSnapshot();

      /**PickAddons step */
      await act(() => {
        fireEvent.click(screen.getAllByRole("button")[1]);
      });

      rerender(<MultiStepForm onSubmit={mockOnSubmit} steps={props} />);

      expect(asFragment()).toMatchSnapshot();

      /**Summary step */
      await act(() => {
        fireEvent.click(screen.getAllByRole("button")[1]);
      });

      rerender(<MultiStepForm onSubmit={mockOnSubmit} steps={props} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("correct functionality", () => {
    it("should submit correct form", async () => {
      /**YourInfo step */
      const { rerender } = render(
        <MultiStepForm onSubmit={mockOnSubmit} steps={props} />
      );

      fireEvent.input(screen.getByRole("textbox", { name: "Name" }), {
        target: { value: "Some name" }
      });
      fireEvent.input(screen.getByRole("textbox", { name: "Email Address" }), {
        target: { value: "some@mail.dom" }
      });
      fireEvent.input(screen.getByRole("textbox", { name: "Phone Number" }), {
        target: { value: "Some address" }
      });

      /**SelectYourPlan step */
      await act(() => {
        fireEvent.click(screen.getByRole("button"));
      });

      rerender(<MultiStepForm onSubmit={mockOnSubmit} steps={props} />);

      const [arcade, advanced, pro] = screen.getAllByRole("radio");

      fireEvent.click(pro);

      expect(arcade).not.toBeChecked();
      expect(advanced).not.toBeChecked();
      expect(pro).toBeChecked();

      await act(async () => {
        fireEvent.click(screen.getByRole("checkbox"));
      });

      /**PickAddons step */
      await act(() => {
        fireEvent.click(screen.getAllByRole("button")[1]);
      });

      rerender(<MultiStepForm onSubmit={mockOnSubmit} steps={props} />);

      fireEvent.click(screen.getByLabelText(/online service/i));
      fireEvent.click(screen.getByLabelText(/larger storage/i));
      fireEvent.click(screen.getByLabelText(/customizable profile/i));

      expect(screen.getByLabelText(/online service/i)).toBeChecked();
      expect(screen.getByLabelText(/larger storage/i)).toBeChecked();
      expect(screen.getByLabelText(/customizable profile/i)).toBeChecked();

      /**Summary step */
      await act(() => {
        fireEvent.click(screen.getAllByRole("button")[1]);
      });

      rerender(<MultiStepForm onSubmit={mockOnSubmit} steps={props} />);

      await act(() => {
        fireEvent.click(screen.getAllByRole("button")[2]);
      });

      expect(formData).toMatchObject({
        name: "Some name",
        email: "some@mail.dom",
        phone: "Some address",
        plan: "pro",
        planMode: false,
        customizableProfile: true,
        onlineService: true,
        largeStorage: true,
        totalPrice: 21
      });
    });
  });
});
