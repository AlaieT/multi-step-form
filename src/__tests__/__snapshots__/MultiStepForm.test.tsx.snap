// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiStepForm correct render should match snapshot for all steps 1`] = `
<DocumentFragment>
  <div
    id="box"
  >
    <div
      id="steps"
    >
      <nav>
        <ul>
          <li
            id="active"
          >
            <p>
              STEP 1
            </p>
            <h3>
              YOUR INFO
            </h3>
          </li>
          <li>
            <p>
              STEP 2
            </p>
            <h3>
              SELECT PLAN
            </h3>
          </li>
          <li>
            <p>
              STEP 3
            </p>
            <h3>
              ADD-ONS
            </h3>
          </li>
          <li>
            <p>
              STEP 4
            </p>
            <h3>
              SUMMARY
            </h3>
          </li>
        </ul>
      </nav>
    </div>
    <form
      aria-label="form"
    >
      <div
        id="step"
      >
        <div
          id="content"
        >
          <div
            id="header"
          >
            <h1>
              Personal info
            </h1>
            <h2>
              Please provide your name, email address, and phone number.
            </h2>
          </div>
          <div
            id="fields"
          >
            <div
              class="input field"
            >
              <div
                class="badges"
              >
                <label
                  for="name"
                >
                  Name
                </label>
              </div>
              <input
                id="name"
                name="name"
                placeholder="e.g. Stephen King"
                type="text"
                value=""
              />
            </div>
            <div
              class="input field"
            >
              <div
                class="badges"
              >
                <label
                  for="email"
                >
                  Email Address
                </label>
              </div>
              <input
                id="email"
                name="email"
                placeholder="e.g. stephenking@lorem.com"
                type="email"
                value=""
              />
            </div>
            <div
              class="input field"
            >
              <div
                class="badges"
              >
                <label
                  for="phone-number"
                >
                  Phone Number
                </label>
              </div>
              <input
                id="phone-number"
                name="phone"
                placeholder="e.q. +1 234 567 890"
                type="tel"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          id="controls"
        >
          <button
            class="button"
            id="next_step"
            type="button"
          >
            Next Step
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`MultiStepForm correct render should match snapshot for all steps 2`] = `
<DocumentFragment>
  <div
    id="box"
  >
    <div
      id="steps"
    >
      <nav>
        <ul>
          <li>
            <p>
              STEP 1
            </p>
            <h3>
              YOUR INFO
            </h3>
          </li>
          <li
            id="active"
          >
            <p>
              STEP 2
            </p>
            <h3>
              SELECT PLAN
            </h3>
          </li>
          <li>
            <p>
              STEP 3
            </p>
            <h3>
              ADD-ONS
            </h3>
          </li>
          <li>
            <p>
              STEP 4
            </p>
            <h3>
              SUMMARY
            </h3>
          </li>
        </ul>
      </nav>
    </div>
    <form
      aria-label="form"
    >
      <div
        id="step"
      >
        <div
          id="content"
        >
          <div
            id="header"
          >
            <h1>
              Select your plan
            </h1>
            <h2>
              You have the option of monthly or yearly billing.
            </h2>
          </div>
          <div
            class="option"
          >
            <input
              checked=""
              id="arcade"
              name="plan"
              type="radio"
              value="arcade"
            />
            <div
              height="40"
              width="40"
            />
            <label
              for="arcade"
            >
              Arcade
            </label>
            <p>
              $90/yr
            </p>
            <p>
              2 months free
            </p>
          </div>
          <div
            class="option"
          >
            <input
              id="advanced"
              name="plan"
              type="radio"
              value="advanced"
            />
            <div
              height="40"
              width="40"
            />
            <label
              for="advanced"
            >
              Advanced
            </label>
            <p>
              $120/yr
            </p>
            <p>
              2 months free
            </p>
          </div>
          <div
            class="option"
          >
            <input
              id="pro"
              name="plan"
              type="radio"
              value="pro"
            />
            <div
              height="40"
              width="40"
            />
            <label
              for="pro"
            >
              Pro
            </label>
            <p>
              $150/yr
            </p>
            <p>
              2 months free
            </p>
          </div>
          <div
            id="switch"
          >
            <label
              for="switch"
            >
              Monthly
            </label>
            <div
              class="switch"
            >
              <input
                checked=""
                id="switch"
                name="planMode"
                type="checkbox"
              />
            </div>
            <label
              for="switch"
              id="selected"
            >
              Yearly
            </label>
          </div>
        </div>
        <div
          id="controls"
        >
          <button
            class="button"
            id="go_back"
            type="button"
          >
            Go Back
          </button>
          <button
            class="button"
            id="next_step"
            type="button"
          >
            Next Step
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`MultiStepForm correct render should match snapshot for all steps 3`] = `
<DocumentFragment>
  <div
    id="box"
  >
    <div
      id="steps"
    >
      <nav>
        <ul>
          <li>
            <p>
              STEP 1
            </p>
            <h3>
              YOUR INFO
            </h3>
          </li>
          <li>
            <p>
              STEP 2
            </p>
            <h3>
              SELECT PLAN
            </h3>
          </li>
          <li
            id="active"
          >
            <p>
              STEP 3
            </p>
            <h3>
              ADD-ONS
            </h3>
          </li>
          <li>
            <p>
              STEP 4
            </p>
            <h3>
              SUMMARY
            </h3>
          </li>
        </ul>
      </nav>
    </div>
    <form
      aria-label="form"
    >
      <div
        id="step"
      >
        <div
          id="content"
        >
          <div
            id="header"
          >
            <h1>
              Pick add-ons
            </h1>
            <h2>
              Add-ons help enhance your gaming experience.
            </h2>
          </div>
          <div
            class="option"
          >
            <input
              id="onlineService"
              name="onlineService"
              type="checkbox"
            />
            <span />
            <div
              class="title"
            >
              <label
                for="onlineService"
              >
                Online service
              </label>
              <label
                for="onlineService"
              >
                Access to multiplayer games
              </label>
            </div>
            +$10/yr
          </div>
          <div
            class="option"
          >
            <input
              id="largeStorage"
              name="largeStorage"
              type="checkbox"
            />
            <span />
            <div
              class="title"
            >
              <label
                for="largeStorage"
              >
                Larger storage
              </label>
              <label
                for="largeStorage"
              >
                Extra 1TB of cloud save
              </label>
            </div>
            +$20/yr
          </div>
          <div
            class="option"
          >
            <input
              id="customizableProfile"
              name="customizableProfile"
              type="checkbox"
            />
            <span />
            <div
              class="title"
            >
              <label
                for="customizableProfile"
              >
                Customizable profile
              </label>
              <label
                for="customizableProfile"
              >
                Custom theme on your profile
              </label>
            </div>
            +$30/yr
          </div>
        </div>
        <div
          id="controls"
        >
          <button
            class="button"
            id="go_back"
            type="button"
          >
            Go Back
          </button>
          <button
            class="button"
            id="next_step"
            type="button"
          >
            Next Step
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`MultiStepForm correct render should match snapshot for all steps 4`] = `
<DocumentFragment>
  <div
    id="box"
  >
    <div
      id="steps"
    >
      <nav>
        <ul>
          <li>
            <p>
              STEP 1
            </p>
            <h3>
              YOUR INFO
            </h3>
          </li>
          <li>
            <p>
              STEP 2
            </p>
            <h3>
              SELECT PLAN
            </h3>
          </li>
          <li>
            <p>
              STEP 3
            </p>
            <h3>
              ADD-ONS
            </h3>
          </li>
          <li
            id="active"
          >
            <p>
              STEP 4
            </p>
            <h3>
              SUMMARY
            </h3>
          </li>
        </ul>
      </nav>
    </div>
    <form
      aria-label="form"
    >
      <div
        id="step"
      >
        <div
          id="content"
        >
          <div
            id="header"
          >
            <h1>
              Finishing up
            </h1>
            <h2>
              Double-check everything looks OK before confirming.
            </h2>
          </div>
          <div
            id="info"
          >
            <div
              id="plan"
            >
              <p>
                arcade yearly
              </p>
              <button
                class="button"
                id="change_plan"
                type="button"
              >
                Change
              </button>
              <p>
                $90/yr
              </p>
            </div>
          </div>
          <div
            id="total"
          >
            <span>
              Total (per year)
            </span>
            <span>
              $90/yr
            </span>
          </div>
        </div>
        <div
          id="controls"
        >
          <button
            class="button"
            id="go_back"
            type="button"
          >
            Go Back
          </button>
          <button
            class="button"
            id="next_step"
            type="submit"
          >
            Confirm
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
